### ISSUE TITLE
YAML indentation error are not detected

### ISSUE THREAD

**Issue by canlar** [dalibo/ldap2pg#203]
Hi,

if i write **managed_roles_query** without space in the yml file i get different result.

correct result: space in (indent)
![lda2pg](https://user-images.githubusercontent.com/18144759/38489654-8e518756-3bd6-11e8-80cd-02b1e526ee9a.JPG) 
Using /var/lib/pgsql/ldap2pg/ldap2pg.yml.
Running in dry mode. Postgres will be untouched.
Inspecting roles in Postgres cluster...
Querying LDAP cn=bs.a.postgresadmins,o... (objectClass...
Role postgresadmins already exists in cluster. Reusing.
Nothing to do.
Comparison complete.

**wrong result without indent in the yml file:**
result give all wrong roles and will delete wrong roles they are NOT superuser. 

![ldap2pg2wrong](https://user-images.githubusercontent.com/18144759/38489891-2714c746-3bd7-11e8-9013-1df4239cbfb8.JPG)

Starting ldap2pg 4.7.
Using /var/lib/pgsql/ldap2pg/ldap2pg.yml.
Running in dry mode. Postgres will be untouched.
Inspecting roles in Postgres cluster...
Querying LDAP cn=bs.a.postgresadmins,o... (objectClass...
Would reassign mysuperuser objects and purge ACL on ICAT.
Would reassign mysuperuser objects and purge ACL on postgres.
Would reassign mysuperuser objects and purge ACL on template1.
Would drop mysuperuser.
Would reassign marketing objects and purge ACL on ICAT.
Would reassign marketing objects and purge ACL on postgres.
Would reassign marketing objects and purge ACL on template1.
Would drop marketing.
Would reassign sales objects and purge ACL on ICAT.
Would reassign sales objects and purge ACL on postgres.
Would reassign sales objects and purge ACL on template1.
Would drop sales.
Would reassign mzmoc objects and purge ACL on ICAT.
Would reassign mzmoc objects and purge ACL on postgres.
Would reassign mzmoc objects and purge ACL on template1.
Would drop mzmoc.
Would reassign group_role objects and purge ACL on ICAT.
Would reassign group_role objects and purge ACL on postgres.
Would reassign group_role objects and purge ACL on template1.
Would drop group_role.
Would reassign doe objects and purge ACL on ICAT.
Would reassign doe objects and purge ACL on postgres.
Would reassign doe objects and purge ACL on template1.
Would drop doe.
Generated 24 querie(s).
Comparison complete.

**Comment by bersace** [comment]
Hallo @canlar .

Why didn't you indent the value ? Did you expect it to work or did you find this in documentation ?

Indentation is meaningful in YAML. If you are not confortable with YAML, you can use plain JSON.

`ldap2pg` is permissive in YAML value to be forward compatible. Maybe a warning would help you to debug such issue ?

**Comment by canlar** [comment]
Hi @bersace,

i can use indent in the file.
I am wondering that i get different result **depends from indent** of the row mamage_roles_query.
second option without indent ignore my select statement and give/drop user that are not superuser.

**Comment by bersace** [comment]
@canlar , i **repeat** :

> Indentation is **meaningful** in YAML.

``` yaml
postgres:
managed_roles_query: whathever
```

Is different from

``` yaml
postgres:
  managed_roles_query: whatever
```

Please keep the issue opened. I will update documentation with hint, warnings and point to YAML documentation.

**Comment by canlar** [comment]
Hi @bersace,

i think there is an error.

If we define only the row managed_roles_query without a select Statement for the whitelist so i get an "Default" execution of select statement for user sync.
all users without superuser role will be deleted,

Same resul(sync/delete all user without superuser role) t you get if i not have  a indent in row managed_roles_query.

I think the default behavior of ldap2pg in case of managed_roles_query must be better checked. Is indent  defined when it must be defined a select statement too. 
Is not a select statement defined when nothing to done.

![managed](https://user-images.githubusercontent.com/18144759/38664756-02892d18-3e2a-11e8-9fb4-fbd48fe44f85.JPG)

**Comment by canlar** [comment]
If i defined empty managed_roles_query: user king will be **_deleted,_**


[ldap2pg.psql         DEBUG] Doing:
[ldap2pg.psql         DEBUG] SELECT
[ldap2pg.psql         DEBUG]   role.rolname, array_agg(members.rolname) AS members, role.rolbypassrls, role.rolcanlogin, role.rolcreatedb, role.rolcreaterole, role.rolinherit, role.rolreplication, role.rolsuper
[ldap2pg.psql         DEBUG] FROM
[ldap2pg.psql         DEBUG]   pg_catalog.pg_roles AS role
[ldap2pg.psql         DEBUG] LEFT JOIN pg_catalog.pg_auth_members ON roleid = role.oid
[ldap2pg.psql         DEBUG] LEFT JOIN pg_catalog.pg_roles AS members ON members.oid = member
[ldap2pg.psql         DEBUG] GROUP BY role.rolname, role.rolbypassrls, role.rolcanlogin, role.rolcreatedb, role.rolcreaterole, role.rolinherit, role.rolreplication, role.rolsuper
[ldap2pg.psql         DEBUG] ORDER BY 1;

_____________________________________________________
[ldap2pg.psql         DEBUG] DROP OWNED BY "king";
[ldap2pg.psql         DEBUG] Connecting to Postgres 'postgres://postgres:admin123@localhost:5432/postgres'.
[ldap2pg.psql          INFO] Would reassign king objects and purge ACL on postgres.
[ldap2pg.psql         DEBUG] Would execute: DO $$BEGIN EXECUTE 'REASSIGN OWNED BY "king" TO '||SESSION_USER; END$$;
[ldap2pg.psql         DEBUG] DROP OWNED BY "king";
[ldap2pg.psql         DEBUG] Connecting to Postgres 'postgres://postgres:admin123@localhost:5432/template1'.
[ldap2pg.psql          INFO] Would reassign king objects and purge ACL on template1.
[ldap2pg.psql         DEBUG] Would execute: DO $$BEGIN EXECUTE 'REASSIGN OWNED BY "king" TO '||SESSION_USER; END$$;
[ldap2pg.psql         DEBUG] DROP OWNED BY "king";
[ldap2pg.psql         DEBUG] Using Postgres connection to 'postgres://postgres:admin123@localhost:5432/'.
[ldap2pg.psql          INFO] Would drop king.
[ldap2pg.psql         DEBUG] Would execute: DROP ROLE "king";
[ldap2pg.manager      DEBUG] No ACL defined. Skipping ACL.
[ldap2pg.manager      DEBUG] Generated 4 querie(s).
[ldap2pg.script        INFO] Comparison complete.

**Comment by bersace** [comment]
Hi @canlar 

By default, `ldap2pg` inspect *all* roles, including superusers.

If `managed_roles_query` is empty or false or none, all users are managed.

What did you expect ?

**Comment by canlar** [comment]
Hi @bersace,

i expect that by "default" ldap2pg don't inpect all roles.
It only inspect only the role member from the ldapgroup "XXXXX" and don't touch other users.

In my test case:
If i have forgot an indent or not define a select statement so all users are removed from database.

Now i know to be sure define a correct ldap group and indent with correct  select statement.

**Comment by bersace** [comment]
> i expect that by "default" ldap2pg don't inpect all roles.

Then what should be the default ? What is the ldapgroup `XXXX` ?

The starting point configuration implements this : https://github.com/dalibo/ldap2pg/blob/master/ldap2pg.yml . It's a configuration not code.

>  If i have forgot an indent or not define a select statement so all users are removed from database.

Yes. This is why ldap2pg run in dry mode by default.

> Now i know to be sure define a correct ldap group and indent with correct select statement.

Yes, indentation is important in YAML, or use `{}` like JSON.

**Comment by bersace** [comment]
@canlar a fix is release with 4.8. Can you test and close the issue ?
